### PR TITLE
Float16 (half or Eigen::half) for conv3d ops

### PR DESCRIPTION
**Pull request by opensourcemattress**
Registrations of conv3d operations with fp16,  fp16 for batch_norms in tf.layers and tf.contrib.layers.  Related issue: https://github.com/tensorflow/tensorflow/issues/11341

I don't understand how fp16 operations must be implemented in low-level with CUDA and I didn't make any additional optimizations.
There is no implementation for fused batch_norm yet.

I copied part of code for dtypes from conv2d test to conv3d test and seems like all works. 
With fp16 I get `inf` and large absolute errors with large numbers. I suppose that it's fine. At least CPU and GPU  implementations return almost similar values. I skip such cases in test.
Example of such case:

> 
> use_gpu: False
> dtype: <dtype: 'float32'>
> data_format: NDHWC
> expected =  [36564.0, 38022.0, 39480.0, 37824.0, 39354.0, 40884.0, 39084.0, 40686.0, 42288.0, 46644.0, 48678.0, 50712.0, 47904.0, 50010.0, 52116.0, 49164.0, 51342.0, 53520.0, 107124.0, 112614.0, 118104.0, 108384.0, 113946.0, 119508.0, 109644.0, 115278.0, 120912.0, 117204.0, 123270.0, 129336.0, 118464.0, 124602.0, 130740.0, 119724.0, 125934.0, 132144.0]
> actual =  [[[[[  36564.   38022.   39480.]
>     [  37824.   39354.   40884.]
>     [  39084.   40686.   42288.]]
> 
>    [[  46644.   48678.   50712.]
>     [  47904.   50010.   52116.]
>     [  49164.   51342.   53520.]]]
> 
> 
>   [[[ 107124.  112614.  118104.]
>     [ 108384.  113946.  119508.]
>     [ 109644.  115278.  120912.]]
> 
>    [[ 117204.  123270.  129336.]
>     [ 118464.  124602.  130740.]
>     [ 119724.  125934.  132144.]]]]]
> 
> use_gpu: False
> dtype: <dtype: 'float16'>
> data_format: NDHWC
> expected =  [36564.0, 38022.0, 39480.0, 37824.0, 39354.0, 40884.0, 39084.0, 40686.0, 42288.0, 46644.0, 48678.0, 50712.0, 47904.0, 50010.0, 52116.0, 49164.0, 51342.0, 53520.0, 107124.0, 112614.0, 118104.0, 108384.0, 113946.0, 119508.0, 109644.0, 115278.0, 120912.0, 117204.0, 123270.0, 129336.0, 118464.0, 124602.0, 130740.0, 119724.0, 125934.0, 132144.0]
> actual =  [[[[[ 36544.  38016.  39488.]
>     [ 37824.  39360.  40896.]
>     [ 39072.  40704.  42304.]]
> 
>    [[ 46656.  48672.  50688.]
>     [ 47936.  50016.  52128.]
>     [ 49184.  51328.  53536.]]]
> 
> 
>   [[[    inf     inf     inf]
>     [    inf     inf     inf]
>     [    inf     inf     inf]]
> 
>    [[    inf     inf     inf]
>     [    inf     inf     inf]
>     [    inf     inf     inf]]]]]
> fp16 using may result in inf values and large absolute errors when used with large numbers, skipping
> 
> use_gpu: True
> dtype: <dtype: 'float32'>
> data_format: NDHWC
> expected =  [36564.0, 38022.0, 39480.0, 37824.0, 39354.0, 40884.0, 39084.0, 40686.0, 42288.0, 46644.0, 48678.0, 50712.0, 47904.0, 50010.0, 52116.0, 49164.0, 51342.0, 53520.0, 107124.0, 112614.0, 118104.0, 108384.0, 113946.0, 119508.0, 109644.0, 115278.0, 120912.0, 117204.0, 123270.0, 129336.0, 118464.0, 124602.0, 130740.0, 119724.0, 125934.0, 132144.0]
> actual =  [[[[[  36564.   38022.   39480.]
>     [  37824.   39354.   40884.]
>     [  39084.   40686.   42288.]]
> 
>    [[  46644.   48678.   50712.]
>     [  47904.   50010.   52116.]
>     [  49164.   51342.   53520.]]]
> 
> 
>   [[[ 107124.  112614.  118104.]
>     [ 108384.  113946.  119508.]
>     [ 109644.  115278.  120912.]]
> 
>    [[ 117204.  123270.  129336.]
>     [ 118464.  124602.  130740.]
>     [ 119724.  125934.  132144.]]]]]
> 
> use_gpu: True
> dtype: <dtype: 'float16'>
> data_format: NDHWC
> expected =  [36564.0, 38022.0, 39480.0, 37824.0, 39354.0, 40884.0, 39084.0, 40686.0, 42288.0, 46644.0, 48678.0, 50712.0, 47904.0, 50010.0, 52116.0, 49164.0, 51342.0, 53520.0, 107124.0, 112614.0, 118104.0, 108384.0, 113946.0, 119508.0, 109644.0, 115278.0, 120912.0, 117204.0, 123270.0, 129336.0, 118464.0, 124602.0, 130740.0, 119724.0, 125934.0, 132144.0]
> actual =  [[[[[ 36576.  38016.  39488.]
>     [ 37824.  39360.  40896.]
>     [ 39072.  40672.  42304.]]
> 
>    [[ 46656.  48672.  50720.]
>     [ 47904.  50016.  52128.]
>     [ 49152.  51328.  53504.]]]
> 
> 
>   [[[    inf     inf     inf]
>     [    inf     inf     inf]
>     [    inf     inf     inf]]
> 
>    [[    inf     inf     inf]
>     [    inf     inf     inf]
>     [    inf     inf     inf]]]]]
> fp16 using may result in inf values and large absolute errors when used with large numbers, skipping

fp16 is not fully covered by tests because I not sure how to do it.

### CODE REVIEW

**Comment by yifeif** [comment]
@yzhwang do you mind taking a look? Thanks!

**Comment by drpngx** [comment]
@yzhwang any cycles for this?

Jenkins, test this please.

**Comment by sb2nov** [comment]
@yzhwang ping

**Comment by yzhwang** [comment]
Sorry I haven't looking at this yet. Will finish it this week.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (yzhwang) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Could you be more specific about large numbers? Or if you could change the value of the input tensors to make them smaller, that would be better. I'm worried that the skipped tests may affect our coverage.

File: tensorflow/python/layers/convolutional.py
Comment (yzhwang) [inline tensorflow/python/layers/convolutional.py]: Check if this breaks backward compatibility please.

File: tensorflow/python/layers/normalization.py
Comment (yzhwang) [inline tensorflow/python/layers/normalization.py]: If you haven't tested this as you do for conv3d, I highly recommend you to split this into another pull request.

**Review by yzhwang — COMMENTED** [review]
Thank you for the pull request. Please address my comments. Also, we are working on enabling half on batch_norm right now. Once the change merges to github, me or my colleague will point you to that change. As for now, I highly recommend you to make this pull request handle conv3d only. It will be much easier to get merged.

File: tensorflow/python/layers/convolutional.py
Comment (opensourcemattress) [reply tensorflow/python/layers/convolutional.py]: I have added this string similarly to conv2d methods.
Tests are rewritten similarly to conv2d tests so compatibility is checked by existing tests. There are both fp16 and fp32 in https://github.com/opensourcemattress/tensorflow/blob/e5f6a30d7b9d95e52fad491b6f951eb71b54df9e/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L56 
and they are used here https://github.com/opensourcemattress/tensorflow/blob/e5f6a30d7b9d95e52fad491b6f951eb71b54df9e/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L94

Comment (opensourcemattress) [reply tensorflow/python/layers/convolutional.py]: I haven't changed this main line https://github.com/opensourcemattress/tensorflow/blob/e5f6a30d7b9d95e52fad491b6f951eb71b54df9e/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L83
It means that there is nothing to do.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: fp16 max value is `65504.0`. So if we try to convolve array and filter which values contain 1e3 it's highly probable that I will got `inf`. Should I write this in code comment?

For now I use test values from old tests for both dtypes and skip only cases with large values. I can try to divide large numbers in fp16 tests. Is it appropriate?

File: tensorflow/core/kernels/conv_grad_ops_3d.cc
Comment (opensourcemattress) [reply tensorflow/core/kernels/conv_grad_ops_3d.cc]: Thanks, I will rewrite it.

File: tensorflow/python/layers/normalization.py
Comment (opensourcemattress) [reply tensorflow/python/layers/normalization.py]: Yes, if you are working on batch norm it's better to remove it from my pull request.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Sorry, I mean "which contain values greater than 1e4", as checked here
 https://github.com/opensourcemattress/tensorflow/blob/0853d075ceaaaf18f5d8de35656635f9e4cd6845/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L122

Comment (yzhwang) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Your explanation here is fine. Please remove the additional logging lines (3 prints) in there (111,112,113). Thank you!

File: tensorflow/python/layers/convolutional.py
Comment (yzhwang) [reply tensorflow/python/layers/convolutional.py]: LGTM.

File: tensorflow/python/layers/normalization.py
Comment (yzhwang) [reply tensorflow/python/layers/normalization.py]: My colleague is working on that. I saw you have removed this. Thank you!

File: tensorflow/core/kernels/conv_grad_ops_3d.cc
Comment (yzhwang) [reply tensorflow/core/kernels/conv_grad_ops_3d.cc]: Thank you.

**Review by yzhwang — COMMENTED** [review]
In general LGTM. Please update the title of your pull request as you've made a few more changes after the initial pull request. Just wait for the test to finish to approve.

File: tensorflow/contrib/layers/python/layers/layers.py
Comment (opensourcemattress) [inline tensorflow/contrib/layers/python/layers/layers.py]: What do you think about this line?
I forgot to remove it in last commit. It adds possibility to use not-fp32-types with tf.contrib.slim.batch_norm and it also doesn't break backward compatibility. 
If you are working on batch norm and it's better to remove it from this pull request, please, add same line in your work.

Comment (yzhwang) [reply tensorflow/contrib/layers/python/layers/layers.py]: @reedwm Would you like to comment on this?

Comment (reedwm) [reply tensorflow/contrib/layers/python/layers/layers.py]: We should not support fp16 and fp64 without tests. Also, for fused_batch_norm, the only fp16 case we support is where the input is fp16 and the scale and offset are in fp32. So if we support fp16 for non-fused batch norm, we should support that case.

In short, I'd remove the line. I plan on adding support for fp16 in layers shortly. If you want to do so, tell me so I don't repeat the work.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Remove the parentheses around data_format, use_gpu.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Before, fp32 rtol was 1e-6, but now its 1e-5. It should be changed back to 1e-6 for fp16.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Please add gradient tests. See the ConstructAndTestGradient function in this file.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: This print statements are unnecessary, so remove them.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: I don't think we have enough tests due to the fact we skip large numbers. For example, testConv3D2x2x2FilterStride2, testConv3DStride3, and testConv3D2x2x2FilterStride2Same are not run at all with fp16. Can you add a few more tests, testing the sorts of things that are currently skipped, that are run with fp16?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Ok. But firstly I need to figure out which absolute errors are appropriate in different cases.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Ok, thanks for mention.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: There is already normalized initial values => small absolute errors. 
I copied code from conv2d tests. 
There was code with float64 in tests but I doubt that it's actually was used (is there fp64 version of conv3d?) so I removed it.

Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Sorry, fp64 was really declared for CPU.

**Comment by opensourcemattress** [comment]
@reedwm, what should I do next?

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Remove the line "orig_strides = strides[:]" above, and change this to "new_strides = test_util.NHWCToNCHW(strides)". Then use new_strides instead of strides.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Remove this print statement.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: As I stated before, this will skip some important tests, causing some things not to be tested. testConv3D2x2x2FilterStride2, testConv3DStride3, and testConv3D2x2x2FilterStride2Same and not run at all with this, so you should add more tests.

@mjanusz, how were the expected outputs calculated for the original tests? It is a pain to add new tests since new expected outputs need to be calculated.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: I made normalization of input and supposed  that current fp64 convs implemented correctly => I took them as reference value. It's not a good idea to test fp64 outputs by themselves but I not sure how to solve this problem in different way. 
It's better just to have reference implementation and just compare outputs but it's additional code and we need more tests for tests, yes? :)

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Can you update this comment? Something like "Initializes the input tensor with array containing numbers from 0 to 1. We keep the input tensor values fairly small to avoid overflowing a float16 tensor during the conv3d."

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Why is this line necessary? Why take the first element for each result?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Done.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Sorry, probably it was used for manual comparison of results. 
I fixed it.

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (reedwm) [inline tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Why does it matter that the absolute error is small? If the absolute error were larger, you can adjust what you pass to atol in self.assertAllClose.

Also, I just realized that you didn't modify what you pass to atol in self.assertAllClose. Can you make it a bit smaller to take into account the expected output values are smaller?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/conv_ops_3d_test.py
Comment (opensourcemattress) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: >Why does it matter that the absolute error is small?

Because if I understand correctly we compare results with reference using absolute difference. And absolute error depend on current significant and exponent. Large numbers (huge exponent) and small amount of bit for significant (fp16) => big absolute errors (I seen difference >10 with huge numbers in some fp16 tests).

I can make 'tolerances' smaller but not sure which values are appropriate. It's hard to tell because convs are not 'pure convs' but some tricky methods from computation mathematics. Error depends on particular implementation of convolution and if we want to be really sure it's better to find out what are relative/absolute errors for particular method and particular type. But I think that from practical point of view it's better to take 'small enough' values and continue to work on other problems.

Comment (reedwm) [reply tensorflow/python/kernel_tests/conv_ops_3d_test.py]: Sorry for late response. I see your point about making 'tolerances' smaller, so you can leave atol as is.

I think you should remove the "and to keep absolute errors small" comment. You're right that having a large exponent with fp16 means large absolute errors, but I don't see that as a problem. We could have just raised 'atol' in that case. Differences greater than 10 seem reasonable given how imprecise fp16 is. However, currently, some expected_outputs are above the maximum fp16 value of 65504, which would cause them to overflow. This seems like the only reason to reduce the numbers in the input tensor IMO.

**Review by reedwm — APPROVED** [review]

**Comment by reedwm** [comment]
Jenkins, test this please.

**Comment by opensourcemattress** [comment]
Thanks for review!

**Comment by reedwm** [comment]
Jenkins, test this please.

**Pull request merged**
